tests/src/test/java/io/streamnative/pulsar/handlers/kop/storage/ProducerStateManagerTest.java

```diff
@@ -0,0 +1,582 @@
+            assertTrue(e.getMessage().contains("which is smaller than the last seen"));
+            log.info("Expected behavior.");
+        }
+    }
+
+    @Test(timeOut = defaultTestTimeout)
+    public void testAppendTxnMarkerWithNoProducerState() {
+        short producerEpoch = 2;
+        appendEndTxnMarker(stateManager, producerId, producerEpoch, ControlRecordType.COMMIT, 27L,
+                -1, time.milliseconds());
+
+        ProducerStateEntry firstEntry = stateManager.lastEntry(producerId).orElseGet(() -> {
+            fail("Expected last entry to be defined");
+            return null;
+        });
+
+        assertEquals(producerEpoch, firstEntry.producerEpoch().shortValue());
+        assertEquals(producerId, firstEntry.producerId());
+        assertEquals(RecordBatch.NO_SEQUENCE, firstEntry.lastSeq().intValue());
+
+        // Fencing should continue to work even if the marker is the only thing left
+        try {
+            append(stateManager, producerId, (short) 0, 0, 0L, 4L, false);
+            fail("Fencing should continue to work even if the marker is the only thing left.");
+        } catch (IllegalArgumentException e) {
+            assertTrue(e.getMessage().contains("which is smaller than the last seen"));
+            log.info("Expected behavior.");
+        }
+
+        // If the transaction marker is the only thing left in the log, then an attempt to write using a non-zero
+        // sequence number should cause an OutOfOrderSequenceException, so that the producer can reset its state
+        try {
+            append(stateManager, producerId, producerEpoch, 17, 0L, 4L, false);
+            fail("If the transaction marker is the only thing left in the log, then an attempt to write"
+                    + "using a non-zero sequence number should cause an OutOfOrderSequenceException, so that the "
+                    + "producer can reset its state.");
+        } catch (OutOfOrderSequenceException e) {
+            log.info("Expected behavior.");
+        }
+
+        // The broker should accept the request if the sequence number is reset to 0
+        append(stateManager, producerId, producerEpoch, 0, 39L, 4L, false);
+        ProducerStateEntry secondEntry = stateManager.lastEntry(producerId).orElseGet(() -> {
+            fail("Expected last entry to be defined");
+            return null;
+        });
+        assertEquals(producerEpoch, secondEntry.producerEpoch().shortValue());
+        assertEquals(producerId, secondEntry.producerId());
+        assertEquals(0, secondEntry.lastSeq().intValue());
+    }
+
+    @Test(timeOut = defaultTestTimeout, expectedExceptions = OutOfOrderSequenceException.class)
+    public void testProducerSequenceInvalidWrapAround() {
+        short epoch = 15;
+        Integer sequence = Integer.MAX_VALUE;
+        long offset = 735L;
+        append(stateManager, producerId, epoch, sequence, offset);
+        append(stateManager, producerId, epoch, 1, offset + 500);
```

Comment:
I found the original test in Kafka is

```scala
    append(stateManager, producerId, epoch, sequence, offset, origin = AppendOrigin.Replication)
    assertThrows(classOf[OutOfOrderSequenceException], () => append(stateManager, producerId, epoch, 1, offset + 500)
```

The first `append` should not throw exceptions. Could you remove `expectedExceptions` and use `assertThrows` instead?